modules/ipc/methods/base.js

```diff
@@ -14,38 +16,101 @@ module.exports = class BaseProcessor {
     }
 
     /**
      * Execute given request.
      * @param  {Object} conn    IPCProviderBackend connection data.
-     * @param  {Object} payload JSON payload object.
+     * @param  {Object|Array} payload JSON payload object
      * @return {Promise}
      */
     exec (conn, payload) {
         this._log.trace('Execute request', payload);
- 
-        return conn.socket.send(payload, {
-            fullResult: true,
+
+        const isBatch = _.isArray(payload);
+
+        const payloadList = isBatch ? payload : [payload];
+
+        // filter out payloads which already have an error
+        const finalPayload = _.filter(payloadList, (p) => {
+            return !p.error;
+        });
+
+        return Q.try(() => {
+            if (finalPayload.length) {
+                return conn.socket.send(finalPayload, {
+                    fullResult: true,
+                });
+            } else {
+                return [];
+            }
         })
-        .then((result) => {
-            /*
-            Result may be a single response or an array of responses.
-             */
-            
-            let resultArray = [].concat(result);
-            let ret = [];
-
-            for (let r of resultArray) {
-                if (r.error) {
-                    throw r.error;
+        .then((ret) => {
+            let result = [];
+
+            _.each(payloadList, (p) => {
+                if (p.error) {
+                    result.push(p);
                 } else {
-                    ret.push(r.result);
+                    p = _.extend({}, p, ret.result.shift());
+
+                    this.sanitizePayload(conn, p);
+
+                    result.push(p);
+                }
+            });
+
+            // if single payload
+            if (!isBatch) {
+                result = result[0];
+
+                // throw error if found
+                if (result.error) {
+                    throw result.error;
                 }
             }
 
-            return _.isArray(result) ? ret : ret[0];
+            return result;
         });
     }
+
+
+    _isAdminConnection (conn) {
+        // main window or popupwindows - always allow requests
+        let wnd = Windows.getById(conn.id);
+
+        return (wnd && ('main' === wnd.type || wnd.isPopup));
+    }
+
+
+    /**
+    Sanitize a request or response payload.
+
+    This will modify the input payload object.
+
+    @param {Object} conn The connection.
+    @param {Object} payload The request payload.
+    */
+    sanitizePayload (conn, payload) {
+        this._log.trace('Sanitize payload', payload);
+
+        if (!_.isObject(payload)) {
+            throw this.ERRORS.INVALID_PAYLOAD;
+        }
+
+        if (this._isAdminConnection(conn)) {
+            return;
+        }
+
+        // prevent dapps from acccesing admin endpoints
+        if(!/^eth_|^shh_|^net_|^web3_|^db_/.test(payload.method)){
+            delete payload.result;
```

Comment:
I add an `error` in the next line. Is deleting the `method` in case of error strictly necessary? I've assumed not.